other/text_to_speech.py

```diff
@@ -0,0 +1,119 @@
+
+Example:
+    text_to_speech(text="¡Hola Mundo!", language="Spanish", voice="Mateo")
+"""
+
+import re
+import subprocess
+import urllib.parse
+
+import requests
+
+LANGUAGES_VOICES = {
+    "US English": ["Alice", "Daisy", "George", "Jenna", "John"],
+    "British": ["Emma", "Harry"],
+    "French": ["Gabriel", "Jade"],
+    "Spanish": ["Isabella", "Mateo"],
+    "German": ["Michael", "Nadine"],
+    "Italian": ["Alessandra", "Giovanni"],
+    "Portuguese": ["Rodrigo"],
+    "Russian": ["Valentina"],
+}
+
+VOICES = {
+    "Emma": "IVONA Amy22 (UK English)",
+    "Harry": "IVONA Brian22 (UK English)",
+    "Jade": "IVONA CΘline22 (French)",
+    "Gabriel": "IVONA Mathieu22 (French)",
+    "Nadine": "IVONA Marlene22 (German)",
+    "Michael": "IVONA Hans22 (German)",
+    "Valentina": "IVONA Tatyana22 (Russian)",
+    "John": "IVONA Eric22",
+    "Jenna": "IVONA Jennifer22",
+    "George": "IVONA Joey22",
+    "Alice": "IVONA Kimberly22",
+    "Daisy": "IVONA Salli22",
+    "Alessandra": "IVONA Carla22 (Italian)",
+    "Giovanni": "IVONA Giorgio22 (Italian)",
+    "Isabella": "IVONA Conchita22 (Spanish [Modern])",
+    "Mateo": "IVONA Enrique22 (Spanish [Modern])",
+    "Rodrigo": "IVONA Cristiano22 (Portuguese)",
+}
+
+
+def text_to_speech(
+    text: str = "Hello world",
+    language: str = "British",
+    voice: str = "Emma",
+    audio_path: str = "output.mp3",
+) -> bool:
+
+    if language not in LANGUAGES_VOICES:
+        raise ValueError(f"Language '{language}' not available")
+
+    if voice not in LANGUAGES_VOICES[language]:
+        raise ValueError(f"Voice '{voice}' not available")
+
+    # prepare data
+    text_encoded = urllib.parse.quote(text.encode("utf8"))
+    data = f"input_text={text_encoded}&language={language}\
+            &voice={VOICES[voice]}&speed=0&action=process_text"
+
+    # make request to www.fromtexttospeech.com
+    # it gives an html with the mp3's url
+    args = [
+        "curl",
+        "http://www.fromtexttospeech.com/",
+        "-H",
+        "Accept-Language: en-US,en;q=0.5",
+        "-H",
+        "Upgrade-Insecure-Requests: 1",
+        "-H",
+        "User-Agent:  \
+            Mozilla/5.0 (X11; Linux x86_64; rv:66.0) Gecko/20100101 Firefox/66.0",
+        "-H",
+        "Content-Type: application/x-www-form-urlencoded",
+        "-H",
+        "Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
+        "-H",
+        "Referer: http://www.fromtexttospeech.com/",
+        "-H",
+        "Connection: keep-alive",
+        "--data",
+        data,
+        "--compressed",
+        "--connect-timeout",
+        "30",
+        "--retry",
+        "300",
+        "--retry-delay",
+        "5",
+        "--max-time",
+        "120",
+    ]
+    process = subprocess.Popen(args, stdout=subprocess.PIPE)
+
+    output = process.communicate()[0]
+
+    # extract the mp3's url
+    mp3_file = re.search(r"(\/output\/\d+\/\d+\.mp3)", output.decode("utf-8")).group(0)
```

Comment:
Please include additional tests for the regex pattern.

Reply:
> Please include additional tests for the regex pattern.

Can you give me a hand? 
How I can only add tests for only that part of the code? And not the entire function?

Reply:
Take a look at the Python `re` module: https://docs.python.org/3/library/re.html#module-re

You compile your regex pattern first using [this method](https://docs.python.org/3/library/re.html#re.compile) and then you can use that as usual but instead of doing `re.search()` you would do `pattern.search()` where `pattern` is the variable name assigned to the [compiled pattern object](https://docs.python.org/3/library/re.html#regular-expression-objects).

This will separate out the pattern object which you can use to test ***only*** the pattern.